src/olaf/simplpedpop.rs

```diff
@@ -0,0 +1,976 @@
+        transcript.append_message(b"context", context);
+
+        let mut bytes = [0; 12];
+        transcript.challenge_bytes(b"nonce", &mut bytes);
+
+        let cipher: ChaCha20Poly1305 = make_aead::<Transcript, ChaCha20Poly1305>(transcript);
+        let nonce = Nonce::from_slice(&bytes[..]);
+
+        let ciphertext: Vec<u8> = cipher
+            .encrypt(nonce, &self.0.as_bytes()[..])
+            .map_err(DKGError::EncryptionError)?;
+
+        Ok(EncryptedSecretShare(ciphertext))
+    }
+}
+
+/// An encrypted secret share, which can be sent directly to the intended participant or through an untrusted coordinator.
+#[derive(Debug, Clone)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct EncryptedSecretShare(pub(crate) Vec<u8>);
+
+impl EncryptedSecretShare {
+    pub(crate) fn decrypt(
+        &self,
+        decryption_key: &Scalar,
+        encryption_key: &RistrettoPoint,
+        context: &[u8],
+    ) -> DKGResult<SecretShare> {
+        let shared_secret = decryption_key * encryption_key;
+
+        let mut transcript = Transcript::new(b"encryption");
+        transcript.commit_point(b"shared secret", &shared_secret.compress());
+        transcript.append_message(b"context", context);
+
+        let mut bytes = [0; 12];
+        transcript.challenge_bytes(b"nonce", &mut bytes);
+
+        let cipher: ChaCha20Poly1305 = make_aead::<Transcript, ChaCha20Poly1305>(transcript);
+        let nonce = Nonce::from_slice(&bytes[..]);
+
+        let plaintext = cipher
+            .decrypt(nonce, &self.0[..])
+            .map_err(DKGError::DecryptionError)?;
+
+        let mut bytes = [0; 32];
+        bytes.copy_from_slice(&plaintext);
+
+        Ok(SecretShare(Scalar::from_bytes_mod_order(bytes)))
+    }
+}
+
+/// SimplPedPoP round 1.
+///
+/// The participant commits to a secret polynomial f(x) of degree t-1, where t is the threshold of the DKG, by commiting
+/// to each one of the t coefficients of the secret polynomial.
+///
+/// It derives a secret key from the secret of the polynomial f(0) and uses it to generate a Proof of Possession of that
+/// secret by signing a message with the secret key.
+pub mod round1 {
+    use super::{
+        derive_secret_key_from_secret, Parameters, ProofOfPossession, SecretPolynomial,
+        SecretPolynomialCommitment,
+    };
+    use crate::{
+        olaf::errors::DKGResult,
+        olaf::polynomial::{Polynomial, PolynomialCommitment},
+        PublicKey, SecretKey,
+    };
+    use core::cmp::Ordering;
+    use curve25519_dalek::Scalar;
+    use merlin::Transcript;
+    use rand_core::{CryptoRng, RngCore};
+
+    /// The private data generated by the participant in round 1.
+    #[derive(Debug, Clone)]
+    #[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+    pub struct PrivateData {
+        pub(crate) secret_key: SecretKey,
+        pub(crate) secret_polynomial: SecretPolynomial,
+    }
```

Comment:
I mentioned session types way back, meaning this struct should not be clone or serialization, and it shiould be consumed by value later.  It's fine if a specific signer device must change this code themselves.

It's also possible it doesn't matter enough here.

Reply:
Yeah, I added because I needed them for the tests and benchmarks...Maybe we can add them only as an internal feature?

Reply:
In fact, we could immediately trash this data if we fix the encryption and make a two phase sender & reciever thing like I discuss elsewhere.